From 7f972278d0f228a7df259867fd3c2b575a59f3bb Mon Sep 17 00:00:00 2001
From: Zhongteng Gui <dragon-archer@outlook.com>
Date: Fri, 16 Oct 2026 22:07:22 +0800
Subject: [PATCH] c++tools: #include <cstdlib> for malloc & free

On platforms where malloc & free aren't implicitly brought in through
other headers (e.g. building with Clang for x86_64-w64-mingw32), this
triggers the following errors:

	../../gcc-16.2.0/c++tools/resolver.cc:124:12: error: use of undeclared identifier 'malloc'
	124 |   buffer = xmalloc (stat.st_size);
	|            ^~~~~~~~~~~~~~~~~~~~~~
	../../gcc-16.2.0/c++tools/resolver.cc:55:20: note: expanded from macro 'xmalloc'
	55 | #define xmalloc(X) malloc(X)
	|                    ^~~~~~
	../../gcc-16.2.0/c++tools/resolver.cc:127:53: error: use of undeclared identifier 'free'
	127 |   struct Deleter { void operator()(void* p) const { free(p); } };
	|                                                     ^~~~
	2 errors generated.

Fix this by directly #include <cstdlib>

c++tools/ChangeLog:

	* resolver.cc: Directly #include <cstdlib> for malloc & free.

Signed-off-by: Zhongteng Gui <dragon-archer@outlook.com>
---
 c++tools/resolver.cc | 1 +
 1 file changed, 1 insertion(+)

diff --git a/c++tools/resolver.cc b/c++tools/resolver.cc
index 53c48b4fc6a..8fa7c49a99b 100644
--- a/c++tools/resolver.cc
+++ b/c++tools/resolver.cc
@@ -25,6 +25,7 @@ along with GCC; see the file COPYING3.  If not see
 #include <algorithm>
 #include <memory>
 // C
+#include <cstdlib> // for malloc & free
 #include <cstring>
 // OS
 #include <fcntl.h>
-- 
2.56.0.windows.2

